### ISSUE TITLE
ePass2003 ATR different, minidriver not working?

### ISSUE THREAD

**Issue by twisteroidambassador** [OpenSC/OpenSC#2572]
### Problem Description

I initialized an ePass2003 on Linux with OpenSC master, and stored a keypair + a certificate  on it. Moving it to a Windows 10 machine with OpenSC 0.22 installed, I could access the token via PKCS11, but not CAPI.

The output of `certutil -scinfo` is :

    > certutil -scinfo
    The Microsoft Smart Card Resource Manager is running.
    Current reader/card status:
    Readers: 1
      0: FS USB Token 0
    --- Reader: FS USB Token 0
    --- Status: SCARD_STATE_PRESENT | SCARD_STATE_INUSE
    --- Status: The card is being shared by a process.
    ---   Card:
    ---    ATR:
            3b 9f 95 81 31 fe 9f 00  66 46 53 05 23 00 25 71   ;...1...fFS.#.%q
            df 00 00 03 90 00 96                               .......
    
    
    =======================================================
    Analyzing card in reader: FS USB Token 0
    SCardGetCardTypeProviderName: The system cannot find the file specified. 0x2 (WIN32: 2 ERROR_FILE_NOT_FOUND)
    Cannot retrieve Provider Name for SCardGetCardTypeProviderName: The system cannot find the file specified. 0x2 (WIN32: 2 ERROR_FILE_NOT_FOUND)
    Cannot retrieve Provider Name for
    --------------===========================--------------
    CertUtil: -SCInfo command FAILED: 0x2 (WIN32: 2 ERROR_FILE_NOT_FOUND)
    CertUtil: The system cannot find the file specified.

So I looked at the registry:

    Windows Registry Editor Version 5.00

    [HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Cryptography\Calais\SmartCards\ePass2003]
    "Crypto Provider"="OpenSC CSP"
    "Smart Card Key Storage Provider"="Microsoft Smart Card Key Storage Provider"
    "80000001"="C:\\Program Files\\OpenSC Project\\OpenSC\\minidriver\\opensc-minidriver.dll"
    "ATR"=hex:3b,9f,95,81,31,fe,9f,00,66,46,53,05,01,00,11,71,df,00,00,03,6a,82,f8
    "ATRMask"=hex:ff,ff,ff,ff,ff,ff,ff,ff,ff,ff,ff,ff,ff,ff,ff,ff,ff,ff,ff,ff,ff,\
      ff,ff

Here I could see that the ATR registered with the driver is different from my token's ATR, and with an ATRMask of all 1's, the driver does not pick up my token at all.

So I compared both ATRs and changed my registry like this:

    "ATR"=hex:3b,9f,95,81,31,fe,9f,00,66,46,53,05,00,00,00,71,df,00,00,03,00,00,00
    "ATRMask"=hex:ff,ff,ff,ff,ff,ff,ff,ff,ff,ff,ff,ff,00,ff,00,ff,ff,ff,ff,ff,00,\
      00,00

With this change, now `certutil -scinfo` can return the contents of the token, and applications using CAPI can access the token as well.

For refrerence, on a different Windows 10 computer I installed Feitian's proprietary driver and tools for ePass2003, and initialized a token. The ATR for that token is:

    ---    ATR:
            3b 9f 95 81 31 fe 9f 00  66 46 53 05 23 00 25 71   ;...1...fFS.#.%q
            df 00 00 00 00 00 05                               .......

And in the registry, the proprietary driver's ATR is:

    Windows Registry Editor Version 5.00
    
    [HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Cryptography\Calais\SmartCards\ePass2003]
    "Crypto Provider"="EnterSafe ePass2003 CSP v1.0"
    "ATR"=hex:3b,9f,95,81,31,fe,9f,00,66,46,53,05,00,00,00,71,df,00,00,00,00,00,00
    "ATRMask"=hex:ff,ff,ff,ff,ff,ff,ff,ff,ff,ff,ff,ff,00,00,00,ff,ff,ff,ff,ff,ff,\
      ff,00

### Proposed Resolution

Figure out why the ATR registered by OpenSC is different from the actual ATR of my token, and why ATR of OpenSC-initialized token is different from proprietary-initialized one;

Determine which ePass2003 tokens the OpenSC driver should support; (for example, should OpenSC minidriver try to interact with a proprietary-initialized token?) 

Change ATR and ATRMask registered by OpenSC driver, so it picks up only those tokens the driver should support.

**Comment by dengert** [comment]
Well I have one of the old epass2003 obtained in 2011 from Gooze.  

The ATR is different then yours:
```
C:\Users\Doug>certutil -scinfo
The Microsoft Smart Card Resource Manager is running.
Current reader/card status:
Readers: 1
  0: FS USB Token 0
--- Reader: FS USB Token 0
--- Status: SCARD_STATE_PRESENT | SCARD_STATE_UNPOWERED
--- Status: The card is available for use.
---   Card:
---    ATR:
        3b 9f 95 81 31 fe 9f 00  66 46 53 05 01 00 11 71   ;...1...fFS....q
        df 00 00 03 90 00 80                               .......


=======================================================
Analyzing card in reader: FS USB Token 0
SCardGetCardTypeProviderName: The system cannot find the file specified. 0x2 (WIN32: 2 ERROR_FILE_NOT_FOUND)
Cannot retrieve Provider Name for SCardGetCardTypeProviderName: The system cannot find the file specified. 0x2 (WIN32: 2 ERROR_FILE_NOT_FOUND)
Cannot retrieve Provider Name for
--------------===========================--------------
CertUtil: -SCInfo command FAILED: 0x2 (WIN32: 2 ERROR_FILE_NOT_FOUND)
CertUtil: The system cannot find the file specified.
```
And the above also fails. Rregistry is pointing at vendor provider not OpenSC.  IIRC when OpenSC is installed and vendor provided entery is in registry, it will not install the OpenSC registry entries. 
```
Windows Registry Editor Version 5.00

[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Cryptography\Calais\SmartCards\ePass2003]
"ATR"=hex:3b,9f,95,81,31,fe,9f,00,66,46,53,05,00,00,00,71,df,00,00,00,00,00,00
"ATRMask"=hex:ff,ff,ff,ff,ff,ff,ff,ff,ff,ff,ff,ff,00,00,00,ff,ff,ff,ff,ff,ff,\
  ff,00
"Crypto Provider"="Microsoft Base Smart Card Crypto Provider"
"Smart Card Key Storage Provider"="Microsoft Smart Card Key Storage Provider"
"80000001"="eps2003csp11.dll"
```

Some things to look at: https://smartcard-atr.apdu.fr/
with your ATR it says 
```
3B 9F 95 81 31 FE 9F 00 66 46 53 05 23 00 25 71 DF 00 00 03 90 00 96

Feitian USB Cryptographic token (FIPS 140-2 Level 3) (PKI)
```

With my old token it  says:
```
3B 9F 95 81 31 FE 9F 00 66 46 53 05 01 00 11 71 DF 00 00 .. .. .. ..

Feitian ePass2003 token
```
The OpenSC driver has an  ATR and mask:
https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/card-epass2003.c#L50-L56

I will leave it up to you to look closer if the above will allow your ATR to be recognized. It could have multiple ATRs listed so it could differentiate between old and new devices.  

Please fell free to submit a PR.

**Comment by twisteroidambassador** [comment]
So far we have collected the following:

    index                           00 01 02 03 04 05 06 07 08 09 0a 0b 0c 0d 0e 0f 10 11 12 13 14 15 16
    -------------------------------+-----------+-----------+-----------+-----------+-----------+--------
    card-epass2003.c ATRMask        FF FF FF FF FF 00 FF FF FF FF FF FF 00 00 00 ff 00 ff ff 00 00 00 00
    card-epass2003.c ATR            3B 9F 95 81 31 FE 9F 00 66 46 53 05 10 00 11 71 df 00 00 00 6a 82 5e
    ePass2003 CSP1.0 ATRMask        ff ff ff ff ff ff ff ff ff ff ff ff 00 00 00 ff ff ff ff ff ff ff 00
    ePass2003 CSP1.0 ATR            3b 9f 95 81 31 fe 9f 00 66 46 53 05 00 00 00 71 df 00 00 00 00 00 00
    Microsoft CSP ATRMask           ff ff ff ff ff ff ff ff ff ff ff ff 00 00 00 ff ff ff ff ff ff ff 00
    Microsoft CSP ATR               3b 9f 95 81 31 fe 9f 00 66 46 53 05 00 00 00 71 df 00 00 00 00 00 00
    my OpenSC-init ATR              3b 9f 95 81 31 fe 9f 00 66 46 53 05 23 00 25 71 df 00 00 03 90 00 96
    my Feitian-init ATR             3b 9f 95 81 31 fe 9f 00 66 46 53 05 23 00 25 71 df 00 00 00 00 00 05
    OpenSC Installed ATR            3b 9f 95 81 31 fe 9f 00 66 46 53 05 01 00 11 71 df 00 00 03 6a 82 f8
    dengert's ATR                   3b 9f 95 81 31 fe 9f 00 66 46 53 05 01 00 11 71 df 00 00 03 90 00 80

The ATR and ATRMask provided here https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/card-epass2003.c#L50-L56 does cover all ePass2003 here, including those initialized by proprietary Feitian tools and therefore incompatible with OpenSC. The ATRMask has a few curious holes at 0x05 and 0x10, which are all consistent across all tokens, so no idea what's going on there.

All OpenSC-initialized tokens have byte 0x13 == 0x03, which means "LCS (life card cycle): 3 (Initialisation state)". Seems like a weird state for the token to be in? Is OpenSC code responsible for setting its value? Can we assume all ePass2003 tokens initialized with OpenSC will have byte 0x13 == 0x03? (I assume **dengert**'s token is also initialized with OpenSC.)

**Comment by dengert** [comment]
`ePass2003 CSP1.0 ATRMask        ff ff ff ff ff ff ff ff ff ff ff ff 00 00 00 ff ff ff ff ff ff ff 00` implies the Feitian minidriver is looking at the "Mandatory status indicator (3 last bytes)" with the LCS and SW and the 
`ePass2003 CSP1.0 ATR            3b 9f 95 81 31 fe 9f 00 66 46 53 05 00 00 00 71 df 00 00 00 00 00 00` implies it expects them to be `00 00 00` 
 
 But the OpenSC code does not care what they are and does not change them from an uninitiated token. 

OpenSC in the [pkcs15init routines ](https://github.com/OpenSC/OpenSC/tree/master/src/pkcs15init) a number of "pkcs15-\<card\>.c" routines call [sc_pkcs15init_set_lifecycle](https://github.com/OpenSC/OpenSC/blob/master/src/pkcs15init/pkcs15-lib.c#L495-L502)
  but  `pkcs15-epass2003.c` does not. So this may be the problem why the Feitian minidriver can not use an OpenSC initialized token. 

This also require the libopensc/card-epass2003.c to implement setting it. via the SC_CARDCTRL_LIFECYCLE.

**Comment by dengert** [comment]
The message `SCardGetCardTypeProviderName: The system cannot find the file specified. 0x2 (WIN32: 2 ERROR_FILE_NOT_FOUND)` may be caused by not finding `eps2003csp11.dll`.

In you registry entry I do not see `"80000001"="eps2003csp11.dll"`  but I do see it in my registry entry. 
A search of my "C:\\" does not find `eps2003csp11.dll`. 

Does your registry entry have `"80000001"="eps2003csp11.dll"`?
Can you find `"80000001"="eps2003csp11.dll"`?

**Comment by dengert** [comment]
Using Windows device manager->Smart cards, I deleted the epass2003 entry. Then after inserting the token a new driver was installed and Driver Details shows it installed `C:\windows\System32\eps2003csp11.dll` and `C:\windows\syswow64\eps2003csp11.dll`   My search now finds these and the ini file. Why they were not found before is not clear. 

But I still get the error. Will look closer next week.

This shows how at the ATR is mapped to a driver:
https://docs.microsoft.com/en-us/windows-hardware/drivers/smartcard/smart-card-plug-and-play

**Comment by twisteroidambassador** [comment]
On the Windows machine with proprietary drivers, the registry indeed does not have a "80000001" value. `certutil -scinfo` can successfully list certificates in the token (that was initialized with Feitian tools), but it does say `SCardGetCardTypeProviderName: The system cannot find the file specified. 0x2 (WIN32: 2 ERROR_FILE_NOT_FOUND)`. Otherwise, the token works as expected.

Anyways, it looks like "Mandatory status indicator" is not intentionally set by OpenSC, and we should not use it to differentiate OpenSC-initialized and and Feitian-initialized tokens. (And maybe consider fixing OpenSC so it generates a "more correct" status indicator?)

I propose changing OpenSC's ATR/Mask to the last lines shown below:

    index                           00 01 02 03 04 05 06 07 08 09 0a 0b 0c 0d 0e 0f 10 11 12 13 14 15 16
    -------------------------------+-----------+-----------+-----------+-----------+-----------+--------
    card-epass2003.c ATRMask        FF FF FF FF FF 00 FF FF FF FF FF FF 00 00 00 ff 00 ff ff 00 00 00 00
    card-epass2003.c ATR            3B 9F 95 81 31 FE 9F 00 66 46 53 05 10 00 11 71 df 00 00 00 6a 82 5e
    ePass2003 CSP1.0 ATRMask        ff ff ff ff ff ff ff ff ff ff ff ff 00 00 00 ff ff ff ff ff ff ff 00
    ePass2003 CSP1.0 ATR            3b 9f 95 81 31 fe 9f 00 66 46 53 05 00 00 00 71 df 00 00 00 00 00 00
    Microsoft CSP ATRMask           ff ff ff ff ff ff ff ff ff ff ff ff 00 00 00 ff ff ff ff ff ff ff 00
    Microsoft CSP ATR               3b 9f 95 81 31 fe 9f 00 66 46 53 05 00 00 00 71 df 00 00 00 00 00 00
    my OpenSC-init ATR              3b 9f 95 81 31 fe 9f 00 66 46 53 05 23 00 25 71 df 00 00 03 90 00 96
    my Feitian-init ATR             3b 9f 95 81 31 fe 9f 00 66 46 53 05 23 00 25 71 df 00 00 00 00 00 05
    OpenSC Installed ATR            3b 9f 95 81 31 fe 9f 00 66 46 53 05 01 00 11 71 df 00 00 03 6a 82 f8
    dengert's ATR                   3b 9f 95 81 31 fe 9f 00 66 46 53 05 01 00 11 71 df 00 00 03 90 00 80
    proposed OpenSC ATRMask         ff ff ff ff ff ff ff ff ff ff ff ff 00 00 00 ff ff ff ff 00 00 00 00
    proposed OpenSC ATR             3b 9f 95 81 31 fe 9f 00 66 46 53 05 00 00 00 71 df 00 00 00 00 00 00

Basically, copy from Feitian's drivers, and additionally mask out the 3 bytes for "Mandatory status indicator". This way it should match all ePass2003 tokens including those already initialized using existing OpenSC versions, and those initialized with a hypothetical future OpenSC version that produce different status indicator bytes. It will also match Feitian-initialized tokens and then fail to read it, though.

**Comment by dengert** [comment]
Some questions:
- Are able to build OpenSC from github? 
- Can you  create a Pull Request (PR)?
- Are you using or going to use ECC key/certs?
- Are you doing this for personal use or for smartcard project for you employer?

If not, I can create the PR. The output of the PR also creates a version of the Windows "msi" files that can be used for testing. 

Although a single ATR and AtrMASK could be used for both cards in the registry, the code in card-epass2003.c may want to distinguish between old and new cards using the "23 00 25" for your token vs "10 00 11" for the old tokens like mine.  (The driver appears to support ECDSA, but I don't see where the curve is specified.)

In regards to the registry entries also see: https://github.com/OpenSC/OpenSC/pull/2523

Windows will cache information about certificates, that refers to the card and driver used to find the private key. When changing drivers i.e. OpenSC vs Fetian, you may need to delete this cached data so it will use the right driver. Control Panel->Internet Options->Content->Certificates->Personal lists these and they can be removed. Problem is they don't really say which are from smart cards.

**Comment by dengert** [comment]
@xaqfan 
You have made many changes to card-epass203. Can you comment on this issue?

Do the ATRs of your epass2003 tokens match  the ones I and @[twisteroidambassador](https://github.com/twisteroidambassador)

**Comment by dengert** [comment]
Sorry did not mean to close this issue.

**Comment by twisteroidambassador** [comment]
Sure, I can make a PR. I can probably also set up a Windows compile environment with some effort, but if this repo already have a CI pipeline for PRs, might as well use it.

This is for personal use.

I am not certain whether my token supports EC. Most likely not, since I read that one would have to special order EC-capable tokens. I did try to generate an EC key, and it failed, but I don't remember the exact command or error message now.

**Comment by dengert** [comment]
@twisteroidambassador @xaqfan  Have a look at #2575  For windows you can find the MSI files under "Some checks were not successful" -> "continuous-integration/appveyor/pr — AppVeyor  details"  which leads to: https://ci.appveyor.com/project/LudovicRousseau/opensc/builds/44081212 There you will find 8 jobs, The 5th leads to the win32 version and the 6th leads to the win64 bit version. The MSI files are under the "artifacts" of each job. 

For the last build of this PR these are: 
https://ci.appveyor.com/api/buildjobs/ptdclvvh54dgxtqq/artifacts/OpenSC-0.22.0_win32.msi
https://ci.appveyor.com/api/buildjobs/peo3n4n3n9qjgx4u/artifacts/OpenSC-0.22.0_win64.msi

I first uninstalled the previous OpenSC 32 and 64 bit version. I also uninstalled some Feitian packages too.

My card does not a certificate, only an RSA key. So the certutil  using the minidriver did not get very far, but got further then before. `C:\Program Files\OpenSC Project\OpenSC\tools>pkcs11-tool --test --login` also worked and showed the  low level card-epass2003.c was being called.

**Comment by twisteroidambassador** [comment]
With these installers, all my ePass2003 tokens are recognized by the OpenSC minidriver, so I think the problem on hand is solved.

(On the other hand, CAPI doesn't actually work well, both with official 0.22 + hand-modified ATR and with these installers. I have an RSA keypair and a self-signed certificate in my OpenSC-initialized token. I can add the certificate to Putty-CAC's `pageant` with its PKCS11 support, and SSH successfully with `putty`. When I add the certificate via CAPI, after typing in the PIN `putty` says "Server refused public-key signature despite accepting key!" I still have to figure out whether I did something wrong or is it an OpenSC problem.)

**Comment by Jakuje** [comment]
> Server refused public-key signature despite accepting key!

This sound like a problem elswhere on the road than on the OpenSC side. This means that the key is accepted, but the putty or putty-cac's pageant provided wrong signature (probably SHA-1 based or somehow else inconsistent), which is not acceptable by the server. Check the putty and ssh server logs for more information

**Comment by dengert** [comment]
Google for: "server refused public-key signature despite accepting key"  
(with or without the quotes). Lots of hits going back a few years.

**Comment by dengert** [comment]
This maybe a SHA-256 issue.  Have a look at #2523
This was to address ECC keys but SHA-256 signatures may also be affected. 

Can you please try changing the registry entries in #2537 version of OpenSC: 
HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Cryptography\Calais\SmartCards\ePass2003
HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\Microsoft\Cryptography\Calais\SmartCards\ePass2003

from:
"Crypto Provider"="OpenSC CSP"
to:
"Crypto Provider"="Microsoft Base Smart Card Crypto Provider"

Then reboot windows and try again. 

#2523 has not gotten many comments. I believe the OpenSC minidriver is not used as much as one might expect. Most smartcard/token vendors provide their own minidriver that is installed by plug-and-play and it just works. So no need to to install OpenSC. 

Your case may be another reason to not mess with the "Crypto Provider" entry in the registry as outlines in #2523

**Comment by dengert** [comment]
What versions of Putty  are  you using? 
What version of ssh daemon and on what system?

https://www.chiark.greenend.org.uk/~sgtatham/putty/changes.html says:

These features were new in [0.75](https://www.chiark.greenend.org.uk/~sgtatham/putty/releases/0.75.html) (released 2021-05-08):
Upgraded default SSH key fingerprint format to OpenSSH-style SHA-256.
Upgraded private key file format to PPK3, with improved passphrase hashing and no use of SHA-1.

Latest release: 
 [0.77](https://www.chiark.greenend.org.uk/~sgtatham/putty/releases/0.77.html) (released 2022-05-27)
